File: apps/api/app/schemas/ml_capabilities.py
"""v0.14.11 · 协议级能力目录响应 schema.

承载 `GET /v1/ml-capabilities/protocol` 的 pydantic 模型。与
`services/capability_registry.py` 的 dataclass 一一对应; 前端 codegen 派生类型。
"""

from __future__ import annotations

from pydantic import BaseModel


class SuggestedBackendItem(BaseModel):
    name: str
    repo_url: str
    summary: str
    research_link: str | None = None
    infra: str | None = None  # 与受控 INFRAS 对齐 (pytorch / onnx / ...)
    builtin: bool = False  # True = 平台 docker-compose 自带; False = 外部推荐


class ProtocolTaskItem(BaseModel):
    id: str
    label: str
    summary: str
    default_geometry: list[str]
    default_modalities: list[str]
    typical_models: list[str]
    protocol_notes: str
    suggested_backends: list[SuggestedBackendItem] = []


class ProtocolInfraItem(BaseModel):
    id: str
    label: str
    summary: str


class ProtocolModalityItem(BaseModel):
    id: str
    label: str
    summary: str


class ProtocolGeometryItem(BaseModel):
    id: str
    label: str
    summary: str


class ProtocolPromptItem(BaseModel):
    # v0.18.30 · prompt 受控词表对外暴露 (供前端 codegen + 运行时同源)。
    # requires_input/interactive_route 双维度见 capability_registry.PromptSpec。
    id: str
    label: str
    summary: str
    requires_input: bool
    interactive_route: bool


class ProtocolInputItem(BaseModel):
    # v0.21.0 · supported_inputs 受控词表对外暴露 (供前端 codegen + 判据同源)。
    id: str
    label: str
    summary: str


class InstanceVariantOption(BaseModel):
    """variants axis 内一条选项 (透传自 backend /setup; 用于前端列表按 axis 拆行)."""

    value: str
    label: str | None = None
    vram_gb: float | None = None
    tier: str | None = None
    recommended: bool = False
    note: str | None = None


class InstanceVariantGroup(BaseModel):
    """variants 多轴的单根轴 (序列保持与 backend /setup 顺序一致)."""

    key: str
    title: str | None = None
    description: str | None = None
    variants: list[InstanceVariantOption] = []


class InstanceModelItem(BaseModel):
    """实例级 model 视图 (字段裁剪自协议 v2 ModelCapability, 不暴露 url / params 等运维
    信息; v0.14.12 起补 supported_variants + variant_combinations 让前端能按主轴展开
    「具体模型」行; v0.19.2 WS0 起补 supported_inputs + resource_profile, 与项目级
    /capabilities 对齐, 供全局编排选择器消费投递契约/批量画像)."""

    id: str
    display_name: str
    task: str = "unknown"
    model_family: str | None = None
    infra: str | None = None
    is_interactive: bool = False
    supported_prompts: list[str] = []
    # v0.19.2 WS0 · 一等输入契约 (full_image|crop|bbox_prompt|point_prompt) + 资源画像
    # (device / batchable)。原 /instances 裁掉了二者, 全局编排选择器需要。
    supported_inputs: list[str] = []
    # v0.21.0 · 源阶段默认输入类型; backend 未声明时由 supported_inputs[0] 兜底。
    default_input_type: str | None = None
    resource_profile: dict = {}
    supported_geometric_outputs: list[str] = []
    supported_trackers: list[str] = []
    # v0.21.19 · text-driven tracker (sam3_video) 子集; 前端选中该 tracker 时显 text 框。
    text_driven_trackers: list[str] = []
    # 协议③ · 属性输出类型 + schema 自描述 (含 select options), 供「从 backend 导入属性」.
    output_attribute_types: list[str] = []
    output_attribute_schema: list[dict] = []
    # v0.20.3 · backend 自报的类别清单 (yolo COCO 等)。供前端「从 backend 预填配置」一键
    # 导入类别——此前本字段在 /instances 被裁掉, 类别只能手抄。缺省 [] 向后兼容。
    # 形态与 MLModelCapability.classes (ml_backend.py) 一致: [{index,name}], 不是 string[] ——
    # 此前误标 list[str] 导致任何自报非空 classes 的 backend (yolo/onnxtools) 校验 80+ 个
    # ValidationError, 整条 backend 被 /instances 路由层静默 catch 掉, 能力目录整体丢失该 backend。
    classes: list[dict] = []
    modality: str | None = None
    supported_variants: list[InstanceVariantGroup] = []
    variant_combinations: list[list[str]] = []
    # v0.14.12 · True 表示同 backend 内多个 task 共享同一份物理权重 (如 gsam2 的 SAM2
    # 权重同时被 segmentation / interactive_seg / tracker 用). 前端列表据此把跨 task
    # 的同 variant 合并到一行, 而不是为每个 task 重复显示。yolo 每 task 独立权重
    # (yolov8n-det.pt vs yolov8n-obb.pt), 缺省 False 保持「每 task 一行 + 任务后缀」。
    variants_shared_across_tasks: bool = False
    # v0.14.13 · backend 自报的默认 variant 组合 (dict[axis_key, value]).
    # 前端 VariantSelector 在用户未选时取此作初值; 优先级链:
    # 项目级 project.default_variants[backend_id] > 本字段 > backend 启动 env 默认.
    default_variants: dict[str, str] = {}
    # 协议 v2.2 · 原子 vs 内部编排维度：atom=单次推理原子；composite=内部编排多原子。
    # 缺省 atom 向后兼容。编排下游 stage 据此过滤（只组合 atom）。
    composition: str = "atom"


class CapabilityInstanceItem(BaseModel):
    """平台已知 backend 的实例视图 (env-only 容器 + 项目级注册合并)。

    `backend_id` 是 ml_backend_registry.id, 供全局编排选择器落 pipeline_stages.ml_backend_id。
    `state` 保留 connected/error 等注册表状态; disconnected 已在 service 层过滤。
    `source` = "env_only" 表示 docker-compose 自带或运维 env 配的容器,
              "registered" 表示项目级 ml_backends 表里有记录。
    `name` 是 backend 显示名 (env-only 取 /setup.name, registered 取 ml_backend.name)。
    不暴露 url / project_id / health 指标, 让普通登录用户安全消费。
    """

    backend_id: str
    state: str
    source: str
    name: str
    infra: str = "unknown"
    # v0.14.14 · backend 是否支持 POST /warmup (协议 §4.4); 前端模型市场"⚡ 预热"
    # 按钮据此置灰. 老 backend 缺字段 = False.
    warmup_endpoint: bool = False
    models: list[InstanceModelItem] = []


class CapabilityInstancesResponse(BaseModel):
    instances: list[CapabilityInstanceItem]


class ProtocolCapabilitiesResponse(BaseModel):
    """协议能力目录响应顶层结构。

    `version` 与 ml-backend-protocol 协议版本对齐 (当前 v2); 受控词表的不兼容
    变更才 bump (新增 task 不算)。
    """

    version: str
    tasks: list[ProtocolTaskItem]
    infras: list[ProtocolInfraItem]
    modalities: list[ProtocolModalityItem]
    geometries: list[ProtocolGeometryItem]
    # v0.18.30 · prompt 受控词表 (第五张; 此前仅内部消费, 现对外暴露供前端 codegen)。
    prompts: list[ProtocolPromptItem] = []
    # v0.21.0 · input 受控词表 (第六张; supported_inputs 合法值)。
    inputs: list[ProtocolInputItem] = []

File: apps/web/src/api/mlCapabilities.ts
// v0.14.11 · 协议级能力目录 API (与 ml backend 注册解耦).
// 后端 SSOT: apps/api/app/services/capability_registry.py.
// 端点: GET /v1/ml-capabilities/protocol — 返回 task / infra / modality / geometry
// 四张受控词表 + 每条 task 的人类可读元数据。

import { useQuery } from "@tanstack/react-query";
import { apiClient } from "./client";

export interface SuggestedBackend {
  name: string;
  repo_url: string;
  summary: string;
  research_link: string | null;
  infra: string | null;
  builtin: boolean;
}

export interface ProtocolTask {
  id: string;
  label: string;
  summary: string;
  default_geometry: string[];
  default_modalities: string[];
  typical_models: string[];
  protocol_notes: string;
  suggested_backends: SuggestedBackend[];
}

export interface ProtocolInfra {
  id: string;
  label: string;
  summary: string;
}

export interface ProtocolModality {
  id: string;
  label: string;
  summary: string;
}

export interface ProtocolGeometry {
  id: string;
  label: string;
  summary: string;
}

export interface ProtocolCapabilities {
  version: string;
  tasks: ProtocolTask[];
  infras: ProtocolInfra[];
  modalities: ProtocolModality[];
  geometries: ProtocolGeometry[];
}

// v0.14.11 · 平台已知 backend 实例 (env-only + 项目级注册合并, 与注册解耦)。
// v0.14.12 起 supported_variants + variant_combinations 透传, 让前端模型市场列表
// 能按 axis 拆「具体模型」行 (yolo 的 series×size 受 MODEL_MATRIX 约束)。
export interface InstanceVariantOption {
  value: string;
  label?: string;
  vram_gb?: number;
  tier?: string;
  recommended?: boolean;
  note?: string;
}

export interface InstanceVariantGroup {
  key: string;
  title?: string;
  description?: string;
  variants?: InstanceVariantOption[];
}

export interface CapabilityInstanceModel {
  id: string;
  display_name: string;
  task: string;
  model_family?: string | null;
  // 能力可见性: internal=目录可见但不对外选用; public/缺省=对外开放.
  visibility?: "internal" | "public";
  infra: string | null;
  is_interactive: boolean;
  supported_prompts: string[];
  supported_geometric_outputs: string[];
  supported_trackers: string[];
  modality: string | null;
  supported_variants?: InstanceVariantGroup[];
  variant_combinations?: string[][];
  // v0.14.12 · 跨 task 共享物理权重 (gsam2 SAM2 一份权重服务 seg/iseg/tracker, 是 true;
  // yolo 每 task 独立权重, 是 false). 前端列表行单位据此切换:
  //   true  → 同 (axis_key, value) 跨 task 合并到一行;
  //   false → 每 task 独立一行, 行名加任务后缀 (YOLOv8-OBB / YOLOv8-Seg).
  variants_shared_across_tasks?: boolean;
  // v0.14.13 · backend 自报的默认 variant 组合 (dict[axis_key, value]).
  // 前端 VariantSelector 在用户未选时取此作初值; 优先级:
  // 项目级 project.default_variants[backend_id] > 本字段 > backend 启动 env 默认.
  default_variants?: Record<string, string>;
  // v0.18.0 · backend 自报的输出属性 schema (协议 §3.x). 二阶段 backend (如 onnxtools
  // 车辆属性) 通过此字段声明 /predict 会写入哪些 attributes (vehicle_type / color 等),
  // 含 select options. 前端「从 ML Backend 导入属性」据此一键导入项目 attribute_schema,
  // 免去手抄选项 + key 对齐。老 backend 缺字段 = 无属性输出。
  output_attribute_schema?: OutputAttributeSchemaItem[];
}

// v0.18.0 · backend 自报的单个输出属性字段 (与项目 AttributeField 同构子集).
export interface OutputAttributeSchemaItem {
  key: string;
  label: string;
  type: string; // text/number/boolean/select/multiselect/range
  options?: { value: string; label: string }[];
}

export interface CapabilityInstance {
  source: "env_only" | "registered" | string;
  name: string;
  infra: string;
  // v0.14.14 · backend 自报是否支持 POST /warmup (协议 §4.4); 老 backend 缺字段 = false.
  // 前端模型市场 "⚡ 预热" 按钮据此置灰.
  warmup_endpoint?: boolean;
  models: CapabilityInstanceModel[];
}

export interface CapabilityInstancesResponse {
  instances: CapabilityInstance[];
}

export const mlCapabilitiesApi = {
  getProtocol: () =>
    apiClient.get<ProtocolCapabilities>("/ml-capabilities/protocol"),
  getInstances: () =>
    apiClient.get<CapabilityInstancesResponse>("/ml-capabilities/instances"),
};

export function useProtocolCapabilities() {
  return useQuery({
    queryKey: ["ml-capabilities", "protocol"],
    queryFn: () => mlCapabilitiesApi.getProtocol(),
    staleTime: 5 * 60_000, // SSOT 是协议常量, 5 分钟缓存足够。
    gcTime: 30 * 60_000,
  });
}

export function useCapabilityInstances() {
  return useQuery({
    queryKey: ["ml-capabilities", "instances"],
    queryFn: () => mlCapabilitiesApi.getInstances(),
    staleTime: 30_000, // 实例层会跟 backend 上下线动态变化, 缓存短一些。
    refetchInterval: 60_000,
  });
}
